Remove `AllocLHeap`.

VSadov opened the pull request:
Fixes:https://github.com/dotnet/runtime/issues/13734

The main difference of `AllocLHeap` is that it uses per-heap allocation context instead of per-thread one. There is no advantage in that and it results in allocation under-reporting in `GetAllocatedBytesForCurrentThread`.

It also came up in POH code reviews as unnecessary inconsistency that we do not want to carry forward.

This change unifies to one allocation entry point - `Alloc` and uses flags to indicate special behaviors.

jkotas commented on src/coreclr/src/gc/gcinterface.h:
```diff
@@ -782,7 +779,7 @@ class IGCHeap {
     virtual Object* AllocAlign8(gc_alloc_context* acontext, size_t size, uint32_t flags) = 0;
```
Do we really need AllocAlign8 ?

VSadov replied:
It is substantially different from regular Alloc and needed only on certain platforms. It seems convenient to have implementation in separate methods.

VSadov replied:
For LOH - we could be forced to use LOH by the caller or decide to use it ourselves based on size. So Alloc has to know about large objects anyways.

jkotas replied:
The implementation can be inside a separate method in the GC. That's fine.

I am looking at it from the GC interface definition point of view. We do have `GC_ALLOC_ALIGN8` flag. It is odd that you can call regular `Alloc` for all other combination flags, except when you have this specific flag - you need to call this other method.

VSadov replied:
I see that `GC_ALLOC_ALIGN8` is not used. Perhaps it was meant to be used in Alloc.

Let's unify Alloc and AllocAlign8 at API level.

jkotas replied:
> I see that GC_ALLOC_ALIGN8 is not used. Perhaps it was meant to be used in Alloc.

It has been used in .NET Native / CoreRT that has things more unified, but this unification did not make it into the master GC sources.

VSadov replied:
Removed `AllocAlign8`. It did reduce some duplication.

jkotas commented on src/coreclr/src/gc/gcinterface.h:
```diff
@@ -782,7 +779,7 @@ class IGCHeap {
 
     // This is for the allocator to indicate it's done allocating a large object during a
-    // background GC as the BGC threads also need to walk LOH.
+    // background GC as the BGC threads also need to walk UOH.
```
What does UOH stand for?

VSadov replied:
User's Old Heaps or User-allocated Other Heaps.

The unifying property is that they are not SOH - i.e. they do not participate in generational behavior. Nothing will promote into or out of these.
That implies they are logically in the oldest generation. As long as card tables are one-directional, these will have to be marked together with Gen2.

It was hard to name ;-)

jkotas replied:
It is weird name. It implies that there are non-user heaps or non-user allocated heaps. Do we have such non-user heaps?

> they are logically in the oldest generation

This would lead to a name like "Oldest Generation Heap", OGH.

VSadov replied:
It is mostly an internal term since the concept of "other than SOH" must have a name. We use it in many places - like comments enums, function and type suffixes. 
If there is a definitely better one, we can change to that (not in this PR).

Ideally the new name would abbreviate as UOH, of course - then no need to change anything. :-)

jkotas replied:
We do use these internal terms to explain how the GC works to our customers. So it is somewhat important that it is a good name.

VSadov replied:
I logged an issue to follow up on this - https://github.com/dotnet/runtime/issues/33445

VSadov committed:
The main difference of `AllocLHeap` is that it uses per-heap acontext instead of per-thread. There is no advantage in that and results in allocations under-reported in `GetAllocatedBytesForCurrentThread`

This change unifies to one allocation entry point - `Alloc` (and its `AllocAlign8` variety)

VSadov committed:
Removed AllocAlign8

jkotas commented on src/coreclr/src/vm/gchelpers.cpp:
```diff
@@ -1175,8 +1092,11 @@ OBJECTREF AllocateObject(MethodTable *pMT
+                (GC_ALLOC_ALIGN8_BIAS | GC_ALLOC_ALIGN8) :
+                GC_ALLOC_ALIGN8;
+
+            orObject = (Object *) Alloc(baseSize, flags);
```
You can merge this with the `else` part.

jkotas commented on src/coreclr/src/vm/gchelpers.cpp:
```diff
@@ -754,7 +671,7 @@ OBJECTREF AllocateArrayEx(MethodTable *pArrayMT, INT32 *pArgs, DWORD dwNumArgs,
             // in their headers so the alignment requirements for the header and the payload are the same.
             _ASSERTE(((pArrayMT->GetBaseSize() - SIZEOF_OBJHEADER) & 7) == 0);
-            orArray = (ArrayBase *) AllocAlign8(totalSize, flags);
+            orArray = (ArrayBase*)Alloc(totalSize, flags | GC_ALLOC_ALIGN8);
```
Same here.

jkotas replied:
(and other places)

VSadov replied:
This is about the 3 places where `GC_ALLOC_ALIGN8` is used, right?

jkotas replied:
Right

VSadov committed:
PR feedback - some refactoring to merge duplicate calls to `Alloc`

jkotas commented on src/coreclr/src/vm/gchelpers.cpp:
```diff
@@ -536,9 +440,20 @@ OBJECTREF AllocateSzArray(MethodTable* pArrayMT, INT32 cElements, GC_ALLOC_FLAGS
-#endif  // FEATURE_64BIT_ALIGNMENT
-
         {
+#else   // ! FEATURE_64BIT_ALIGNMENT
```
Two ifdefs may be more readable than #if / #else / #endif. #ifdefs around unmatched `{` are hard to read.

VSadov replied:
It was also causing build issues on arm.

VSadov committed:
Splited an `ifdef/else`  in two to not cross code blocks.

Maoni0 commented on src/coreclr/src/gc/gcinterface.h:
```diff
@@ -7,7 +7,7 @@
 // The major version of the GC/EE interface. Breaking changes to this interface
 // require bumps in the major version number.
-#define GC_INTERFACE_MAJOR_VERSION 4
+#define GC_INTERFACE_MAJOR_VERSION 5
```
I don't think you need to increase this again, it was already increased with your last checkin to this file. we just need to increase it each time we actually release something

VSadov replied:
I see. I thought we literally have to update every time we merge breaking changes :-)

VSadov replied:
Was a bit surprised the number is so low.

Maoni0 replied:
every time we make breaking changes that are publicly visible.

Maoni0 reviewed (approved):
:shipit:

VSadov committed:
No need to update `GC_INTERFACE_MAJOR_VERSION` more than once per release. And we already did for this one.

VSadov commented:
Thanks!!
